src/core/exception.d

```diff
@@ -416,12 +427,65 @@ deprecated void setAssertHandler( AssertHandler h ) @trusted nothrow @nogc
     assertHandler = h;
 }
 
+/*
+Returns a perThreadInstance object of class type `C`, i.e. repeated calls to
+`perThreadInstance!C` within a given thread return the same instance. For that
+reason, no constructor parameters are accepted; `C` must allow default
+construction.
+
+Different threads own different instances of `C`.
+
+The destructor of the perThreadInstance object is called during application's
+exit.
+
+Returns: A thread-local instance of type `C`.
+
+Throws: If the default constructor of `C` throws.
+*/
+private C perThreadInstance(C)() if (is(C == class) && is(typeof(new C)))
+{
+    static size_t[1 + (__traits(classInstanceSize, C) - 1) / size_t.sizeof] b;
+    static bool initialized = false;
+    auto result = () @trusted { return cast(C) b.ptr; }();
+    if (!initialized)
+    {
+        assert(typeid(C).initializer.length == b.length * size_t.sizeof);
+        import core.stdc.string;
+        () @trusted
+        {
+            memcpy(b.ptr, typeid(C).initializer.ptr, b.length);
+        }();
+        static if (is(typeof(result.__ctor())))
+            result.__ctor;
+        import core.internal.traits;
+        static if (hasElaborateDestructor!C)
+        {
+            import core.stdc.stdlib;
+            static void destroy()
+            {
+                (cast(C) b.ptr).__dtor;
```

Comment:
If you don't want to leverage `object.destroy`, you should at least use `rt_finalize2`, because that's what `object.destroy` [does for classes](https://github.com/dlang/druntime/blob/v2.072.1/src/object.d#L2718).
The problem with calling `obj.__dtor` directly is that:
+ It does a non-virtual call (see [1](https://godbolt.org/g/y2zSnr) and [2](https://godbolt.org/g/e7Am3H)) to `C.~this()`, without calling any of `C`'s base classes' dtors.
+ It does not destroy `C`'s `__monitor` field.
+ It does not clear the memory, causing the GC to attempt destroy (this time "properly") the object again - see [this](http://rextester.com/ROLT52146) (reproduces perfectly on newer compilers too).